### PR TITLE
python3Packages.cirq: init at 0.6.1

### PR DESCRIPTION
**Pull request by drewrisinger**
###### Motivation for this change

Add Google's Cirq package.
Uses: Quantum information science.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/cirq/default.nix [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+}:
+
+buildPythonPackage rec {
+  pname = "cirq";
+  version = "0.6.1";
+
+  disabled = pythonOlder "3.5";
+
+  # TODO: replace fetchFromGitHub with fetchPypi in future if don't need patching
+  # Pypi doesn't host source code for Cirq, only wheels, which doesn't work with requirements patch
```
Comment (timokau) [inline pkgs/development/python-modules/cirq/default.nix]: Why would we ever want to replace the source with the wheel then?

File: pkgs/development/python-modules/cirq/default.nix [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    repo = "cirq";
+    rev = "v${version}";
+    sha256 = "0lhr2dka7vpz9xd6akxphrcv2b3ni2cgjywpc1r7qpqa5mrq1q7f";
+  };
+
+  # Cirq 0.6 requires networkx==2.3 only for qiskit dependency/test, force this to avoid networkx version conflicts
+  prePatch = ''
+    substituteInPlace requirements.txt --replace "networkx==2.3" "networkx"
+
+    # Fix sympy 1.5 test failures. Should be fixed in v0.7
```
Comment (timokau) [inline pkgs/development/python-modules/cirq/default.nix]: Is there some commit from 0.7 that we could cherry-pick instead? Adding a patch is nicer than manually patching, since it will fail to apply once it has reached upstream. Random `sed` invocations tend to stick around, as they don't fail when they aren't necessary anymore.

Also an upstream reference would be nice either way.

File: pkgs/development/python-modules/cirq/default.nix [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  # TODO: replace fetchFromGitHub with fetchPypi in future if don't need patching
+  # Pypi doesn't host source code for Cirq, only wheels, which doesn't work with requirements patch
+  src = fetchFromGitHub {
+    owner = "quantumlib";
+    repo = "cirq";
+    rev = "v${version}";
+    sha256 = "0lhr2dka7vpz9xd6akxphrcv2b3ni2cgjywpc1r7qpqa5mrq1q7f";
+  };
+
+  # Cirq 0.6 requires networkx==2.3 only for qiskit dependency/test, force this to avoid networkx version conflicts
```
Comment (timokau) [inline pkgs/development/python-modules/cirq/default.nix]: Is there some upstream issue we could reference here?

File: pkgs/development/python-modules/cirq/default.nix [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    substituteInPlace cirq/contrib/quirk/cells/parse_test.py --replace "parse_formula('5t') == 5 * t" "parse_formula('5t') == 5.0 * t"
+  '';
+
+  propagatedBuildInputs = [
+    google_api_python_client
+    numpy
+    matplotlib
+    networkx
+    pandas
+    (pythonProtobuf.override { protobuf = protobuf; })
```
Comment (timokau) [inline pkgs/development/python-modules/cirq/default.nix]: Why is this necessary?

File: pkgs/development/python-modules/cirq/default.nix [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    requests
+    scipy
+    sortedcontainers
+    sympy
+    typing-extensions
+  ];
+
+  doCheck = true;
+  # pythonImportsCheck = [ "cirq" "cirq.Ciruit" ];  # cirq's importlib hook doesn't work here
```
Comment (timokau) [inline pkgs/development/python-modules/cirq/default.nix]: How/why does it fail?

File: pkgs/development/python-modules/cirq/default.nix [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+  dontUseSetuptoolsCheck = true;
```
Comment (timokau) [inline pkgs/development/python-modules/cirq/default.nix]: Why is this necessary?

File: pkgs/development/python-modules/cirq/default.nix [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+  checkInputs = [
+    pytestCheckHook
+    pytest-benchmark
+    ply
+    pydot
+    pyyaml
+    pygraphviz
+  ];
+  # TODO: enable op_serializer_test. Error is type checking, for some reason wants bool instead of numpy.bool_. Not sure if protobuf or internal issue
```
Comment (timokau) [inline pkgs/development/python-modules/cirq/default.nix]: Is there an upstream issue for that?

Also please add a comment that explains why we ignore `dev_tools`.

File: pkgs/development/python-modules/cirq/default.nix [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+  pytestFlagsArray = [
+    "--ignore=dev_tools"
+    "--ignore=cirq/google/op_serializer_test.py"
+  ];
+
+  meta = with lib; {
+    description = "A framework for creating, editing, and invoking Noisy Intermediate Scale Quantum (NISQ) circuits.";
```
Comment (timokau) [inline pkgs/development/python-modules/cirq/default.nix]: Please strip the period at the end (see https://nixos.org/nixpkgs/manual/#sec-standard-meta-attributes).

File: pkgs/development/python-modules/sympy/default.nix [hunk 9]
```diff
@@ -8,11 +8,11 @@
 
 buildPythonPackage rec {
   pname = "sympy";
-  version = "1.5"; # Upgrades may break sage. Please test or ping @timokau.
+  version = "1.5.1"; # Upgrades may break sage. Please test or ping @timokau.
```
Comment (timokau) [inline pkgs/development/python-modules/sympy/default.nix]: Please rebase on current master.

File: maintainers/maintainer-list.nix [hunk 10]
```diff
@@ -1905,6 +1905,12 @@
     email = "burkett.andrew@gmail.com";
     name = "Andrew Burkett";
   };
+  drewrisinger = {
+    email = "drisinger+nixpkgs@gmail.com";
+    github = "drewrisinger";
+    gitHubId = 10198051;
+    name = "Drew Risinger";
+  };
```
Comment (timokau) [inline maintainers/maintainer-list.nix]: Welcome to the team :)

**Review by timokau — APPROVED** [review]
Looks very good to me! Still, some nitpicks. They are not particularly important, but I tend to keep the bar for new packages quite high.

**Review by timokau — CHANGES_REQUESTED** [review]
Meant to request changes, not approve.

File: pkgs/development/python-modules/cirq/default.nix [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+  src = fetchFromGitHub {
+    owner = "quantumlib";
+    repo = "cirq";
+    rev = "v${version}";
+    sha256 = "0lhr2dka7vpz9xd6akxphrcv2b3ni2cgjywpc1r7qpqa5mrq1q7f";
+  };
+
+  # Cirq 0.6 requires networkx==2.3 only for qiskit dependency/test, force this to avoid networkx version conflicts
+  prePatch = ''
+    substituteInPlace requirements.txt --replace "networkx==2.3" "networkx"
```
Comment (jonringer) [inline pkgs/development/python-modules/cirq/default.nix]: this will cause a failure likely, network==2.4 introduced a regression where they removed `Digraph.node` attribute

File: pkgs/development/python-modules/cirq/default.nix (hunk 1 above)
Comment (drewrisinger) [reply pkgs/development/python-modules/cirq/default.nix]: Fair point.

File: pkgs/development/python-modules/cirq/default.nix (hunk 2 above)
Comment (drewrisinger) [reply pkgs/development/python-modules/cirq/default.nix]: Tried using fetchpatch earlier. Fails b/c the PR with the patch is based on restructuring some of the tests, so files aren't in the right place.

I'll take a quick stab at 0.7, see what happens.

File: pkgs/development/python-modules/cirq/default.nix (hunk 6 above)
Comment (drewrisinger) [reply pkgs/development/python-modules/cirq/default.nix]: if ``dontUseSetuptoolsCheck`` isn't true, then ``setuptools test`` fails (i.e. ``setuptoolsCheckPhase``) b/c it can't find setup.py. ``pytestCheckHook`` will automatically run ``pytest``, so don't need ``setuptools test`` at all.

File: pkgs/development/python-modules/cirq/default.nix (hunk 7 above)
Comment (drewrisinger) [reply pkgs/development/python-modules/cirq/default.nix]: No, I have no idea why op_serializer test is failing. Upstream seems fine.

File: pkgs/development/python-modules/cirq/default.nix (hunk 11 above)
Comment (drewrisinger) [reply pkgs/development/python-modules/cirq/default.nix]: Per https://github.com/quantumlib/Cirq/issues/2368, the only failure is for qiskit, an optional dependency which isn't imported/tested in this build. Build works locally with networkx >= 2.4

File: pkgs/development/python-modules/cirq/default.nix (hunk 4 above)
Comment (drewrisinger) [reply pkgs/development/python-modules/cirq/default.nix]: pythonProtobuf in repo is 3.7.1, but this requires protobuf==3.8.0 (https://github.com/quantumlib/Cirq/blob/0882abb751005191f12d9bae410c9851ea705041/requirements.txt#L9). I didn't know if bumping protobuf would break other packages, so I thought that rebuilding pyProtobuf with protobuf3_8 would be better. It's locked there here (in python-packages.nix): https://github.com/drewrisinger/nixpkgs/blob/512b5d392cbed4a2129f37ba3a40d70e7704b4e8/pkgs/top-level/python-packages.nix#L1825-L1828

I guess other option is bumping pythonPackages.protobuf -> 3.8.0

File: pkgs/development/python-modules/cirq/default.nix (hunk 3 above)
Comment (drewrisinger) [reply pkgs/development/python-modules/cirq/default.nix]: https://github.com/quantumlib/Cirq/issues/2368. Will include in comment

File: pkgs/development/python-modules/cirq/default.nix (hunk 5 above)
Comment (drewrisinger) [reply pkgs/development/python-modules/cirq/default.nix]: When uncommented: 
```bash
Check whether the following modules can be imported: cirq cirq.Ciruit
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "<string>", line 1, in <lambda>
  File "/nix/store/aw85aijllnc5spk2fqbivygrhx6x7411-python3-3.7.6/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/nix/store/j99g9hj4fhqpznbjsd3sqa9dzh5pj1v6-python3.7-cirq-0.6.1/lib/python3.7/site-packages/cirq/__init__.py", line 15, in <module>
    from cirq import _import
  File "/nix/store/j99g9hj4fhqpznbjsd3sqa9dzh5pj1v6-python3.7-cirq-0.6.1/lib/python3.7/site-packages/cirq/_import.py", line 25, in <module>
    class InstrumentedFinder(importlib.abc.MetaPathFinder):
AttributeError: module 'importlib' has no attribute 'abc'
```

Not sure why it can't find this module, ``importlib.abc`` has been around since py3.3 (I think). Only guess is something with eval/freezing imports?

**Comment by drewrisinger** [comment]
Tried with cirq v0.7.0, and got following failures.

```
________________________ test_serialize_sympy_constants ________________________

    def test_serialize_sympy_constants():
        proto = _arg_to_proto(sympy.pi, arg_function_language='')
        packed = json_format.MessageToDict(proto,
                                           including_default_value_fields=True,
                                           preserving_proto_field_name=True,
                                           use_integers_for_enums=True)
>       assert packed == {'arg_value': {'float_value': float(np.float32(sympy.pi))}}
E       AssertionError: assert {'arg_value':...592653589793}} == {'arg_value':...927410125732}}
E         Differing items:
E         {'arg_value': {'float_value': 3.141592653589793}} != {'arg_value': {'float_value': 3.1415927410125732}}
E         Use -v to get the full diff

cirq/google/arg_func_langs_test.py:108: AssertionError
___________________ test_serialize_conversion[value1-proto1] ___________________

value = array([ True, False])
proto = {'arg_value': {'bool_values': {'values': [True, False]}}}

    @pytest.mark.parametrize('value,proto', [
        ((True, False), {
            'arg_value': {
                'bool_values': {
                    'values': [True, False]
                }
            }
        }),
        (np.array([True, False], dtype=np.bool), {
            'arg_value': {
                'bool_values': {
                    'values': [True, False]
                }
            }
        }),
    ])
    def test_serialize_conversion(value: ARG_LIKE, proto: v2.program_pb2.Arg):
        msg = v2.program_pb2.Arg()
        json_format.ParseDict(proto, msg)
        packed = json_format.MessageToDict(_arg_to_proto(value,
>                                                        arg_function_language=''),
                                           including_default_value_fields=True,
                                           preserving_proto_field_name=True,
                                           use_integers_for_enums=True)

cirq/google/arg_func_langs_test.py:138: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
cirq/google/arg_func_langs.py:131: in _arg_to_proto
    msg.arg_value.bool_values.values.extend(value)
/nix/store/86g6marsgjjr7yqs903rzp0dynxbcck0-python3.7-protobuf-3.8.0/lib/python3.7/site-packages/google/protobuf/internal/containers.py:281: in extend
    new_values = [self._type_checker.CheckValue(elem) for elem in elem_seq_iter]
/nix/store/86g6marsgjjr7yqs903rzp0dynxbcck0-python3.7-protobuf-3.8.0/lib/python3.7/site-packages/google/protobuf/internal/containers.py:281: in <listcomp>
    new_values = [self._type_checker.CheckValue(elem) for elem in elem_seq_iter]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <google.protobuf.internal.type_checkers.TypeCheckerWithDefault object at 0x7fffd776ad10>
proposed_value = True

    def CheckValue(self, proposed_value):
      """Type check the provided value and return it.
    
      The returned value might have been normalized to another type.
      """
      if not isinstance(proposed_value, self._acceptable_types):
        message = ('%.1024r has type %s, but expected one of: %s' %
                   (proposed_value, type(proposed_value), self._acceptable_types))
>       raise TypeError(message)
E       TypeError: True has type <class 'numpy.bool_'>, but expected one of: (<class 'bool'>, <class 'numbers.Integral'>)

/nix/store/86g6marsgjjr7yqs903rzp0dynxbcck0-python3.7-protobuf-3.8.0/lib/python3.7/site-packages/google/protobuf/internal/type_checkers.py:109: TypeError
______________________________ test_async_sample _______________________________

pyfuncitem = <Function test_async_sample>

    def pytest_pyfunc_call(pyfuncitem):
        if inspect.iscoroutinefunction(pyfuncitem._obj):
            # coverage: ignore
>           raise ValueError(f'{pyfuncitem._obj.__name__} is async but not '
                             f'decorated with "@pytest.mark.asyncio".')
E           ValueError: test_async_sample is async but not decorated with "@pytest.mark.asyncio".

cirq/conftest.py:14: ValueError
_____________________________ test_asyncio_pending _____________________________
[...] MORE of similar error as above
```

File: pkgs/development/python-modules/cirq/default.nix (hunk 2 above)
Comment (drewrisinger) [reply pkgs/development/python-modules/cirq/default.nix]: Relevant issue: https://github.com/quantumlib/Cirq/pull/2655

File: pkgs/development/python-modules/cirq/default.nix (hunk 4 above)
Comment (timokau) [reply pkgs/development/python-modules/cirq/default.nix]: I think it would be good to at least investigate *why* the main protobuf is still at 3.7. If there's no good reason, better to bump it than to accumulate a bunch of special cases.

File: pkgs/development/python-modules/cirq/default.nix (hunk 5 above)
Comment (timokau) [reply pkgs/development/python-modules/cirq/default.nix]: Hm, not sure what exactly the import check is doing here. I assumed if its importable at runtime, it should be importable here. Weird.

File: pkgs/development/python-modules/cirq/default.nix (hunk 6 above)
Comment (timokau) [reply pkgs/development/python-modules/cirq/default.nix]: Okay, I assumed the `pytestCheckHook` would automatically replace the normal check phase. Apparently not the case.

File: pkgs/development/python-modules/cirq/default.nix (hunk 7 above)
Comment (timokau) [reply pkgs/development/python-modules/cirq/default.nix]: Can you look into it? Maybe opening an issue upstream is still a good idea, they might have an idea what could be going wrong.

File: pkgs/development/python-modules/cirq/default.nix (hunk 4 above)
Comment (drewrisinger) [reply pkgs/development/python-modules/cirq/default.nix]: Fair. But tests pass with default protobuf (3.7), so I'm not going to worry about it. I didn't see any notes in 985b9e47b8f4e96fcf4dc691e69e7813827b7ee5 (last bump) about why it was bumped.

File: pkgs/development/python-modules/cirq/default.nix (hunk 7 above)
Comment (drewrisinger) [reply pkgs/development/python-modules/cirq/default.nix]: Opened: https://github.com/quantumlib/Cirq/issues/2727

File: pkgs/development/python-modules/cirq/default.nix (hunk 5 above)
Comment (drewrisinger) [reply pkgs/development/python-modules/cirq/default.nix]: Am I OK to resolve this? Package should be sufficiently tested that I'm not going to worry about this one failure.

**Comment by timokau** [comment]
Looks good to me now. It would be nice to resolve the outstanding issues and update to the latest release though. Let's give upstream some time to reply to your issue.

If nothing has happened until monday, feel free to ping me again and I'll merge in the current state.

**Review by timokau — APPROVED** [review]

**Comment by timokau** [comment]
Thinking about this some more, I think it can already be pretty useful in its current state and the 0.7 update (and hopefully the fix of the remaining issues) can be done as a follow-up. Thank you!

**Comment by drewrisinger** [comment]
👍
